workflow-manager/task/task.go

```diff
@@ -163,41 +165,45 @@ func NewGCPPubSubEnqueuer(project string, topicID string, dryRun bool) (*GCPPubS
 	}
 
 	return &GCPPubSubEnqueuer{
-		topic:  client.Topic(topicID),
-		dryRun: dryRun,
+		topic:   client.Topic(topicID),
+		dryRun:  dryRun,
+		limiter: limiter.New(maxWorkers),
 	}, nil
 }
 
 func (e *GCPPubSubEnqueuer) Enqueue(task Task, completion func(error)) {
-	e.waitGroup.Add(1)
-	go func(task Task) {
-		defer e.waitGroup.Done()
-		jsonTask, err := json.Marshal(task)
-		if err != nil {
-			completion(fmt.Errorf("marshaling task to JSON: %w", err))
-			return
-		}
-
-		if e.dryRun {
-			log.Printf("dry run, not enqueuing task")
+	e.limiter.Execute(func(ticket *limiter.Ticket) {
+		defer e.limiter.Done(ticket)
```

Comment:
This function (the one passed to `Execute()` will return immediately after creating the goroutine on line 178, and so the ticket will be returned to the limiter before the potentially expensive work is done. I think this should be called inside the closure that is created on line 178.

Reply:
Oh I completely didn't notice the goroutine there. Thanks!